### PR TITLE
ARTEMIS-2212 Avoid using CLQ on ServerConsumerImpl

### PR DESCRIPTION
**Pull request by franz1981**
It would deliver a better performance for the most
common operations eg offer, poll, iterations, size.

### CODE REVIEW

**Comment by franz1981** [comment]
@michaelandrepearce and @clebertsuconic  in the new here please take a look at this one :+1: 
I can change ArrayQueue with a LinkedList and I think that the benefits will be present due to avoiding a concurrent data structure, but we loose the added benefit of avoiding the garbage produced by the `Node` allocated.

**Comment by michaelandrepearce** [comment]
Whats the concurrent performance on this. Do you have some stats?

Re sync lock blocks. Could be replaced with lock on this and or sync methods.

**Comment by franz1981** [comment]
> Whats the concurrent performance on this. Do you have some stats?

Only allocations profiles and given that is in the hot path, we have a LOT of ConcurrentLinkedQueue::Node saved by this change.
The overall perf won't change by a bit (if not a tiny improvement), because it isn't a bottleneck AFAIK.

> Re sync lock blocks. Could be replaced with lock on this and or sync methods.

Agree at a first look, but I do not remember the reason why was used a separate lock TBH

**Comment by michaelandrepearce** [comment]
If its not exposed and its used like doing lock on 'this". Then i can only assume it was the coders pattern at that time.

**Comment by michaelandrepearce** [comment]
@franz1981, did you make the change to syncronized methods? As this PR is about performance im for getting the most we can in one ;)

**Comment by franz1981** [comment]
@michaelandrepearce Not yet, need to chenge it and we just need the opinion from @clebertsuconic here, given that we I'm concerned about the max size that the array q can get and that will be retained for the whole life-cycle of the consumer :)

**Comment by franz1981** [comment]
@michaelandrepearce I'm taking another look and I'm not 100% sure to replace the usage of the lock object with `this` or synchronized methods, because I see that the synchronization on `this` has a different semantic and uses, separated from `lock` eg 
```java
public HandleStatus handle(final MessageReference ref) throws Exception {
      // available credits can be set back to null with a flow control option.
      AtomicInteger checkInteger = availableCredits;
      if (callback != null && !callback.hasCredits(this) || checkInteger != null && checkInteger.get() <= 0) {
         if (logger.isDebugEnabled()) {
            logger.debug(this + " is busy for the lack of credits. Current credits = " +
                            availableCredits +
                            " Can't receive reference " +
                            ref);
         }

         return HandleStatus.BUSY;
      }

      synchronized (lock) {
```
`handle` is not synchronized, while other methods are, but a part of it `synchronized (lock)`, making the behaviour different if I replace `synchronized (lock)` with a `synchronized (this)` ie the risk is to be blocked by other method calls that currently are not blocking it

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java [hunk 1]
```diff
@@ -509,21 +518,16 @@ public synchronized void close(final boolean failed) throws Exception {
...
-
-      Iterator<MessageReference> iter = refs.iterator();
+      List<MessageReference> refs = cancelRefs(failed, false, null);
 
       Transaction tx = new TransactionImpl(storageManager);
 
-      while (iter.hasNext()) {
-         MessageReference ref = iter.next();
-
+      refs.forEach(ref -> {
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java]: is this atomic?

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java]: Ohhh... ok.. it's on the list from cancelRefs.. I confused as being the main list.

**Pull request closed without merging**